lib/internal/fs/promises.js

```diff
@@ -665,9 +665,22 @@ async function mkdir(path, options) {
                        kUsePromises);
 }
 
+const opendir = promisify(_opendir);
+
 async function readdir(path, options) {
   options = getOptions(options, {});
   path = getValidatedPath(path);
+  if (options.recursive) {
+    const dir = await opendir(path, { recursive: true });
+    const dirents = [];
+    for await (const dirent of dir) {
```

Comment:
Where is the async iteration of `dir` implemented?

Reply:
The current `Dir` is an async iterator see: `async* entries()`. The current implementation was already reading the entire directory into memory and returning that; I thought for this first implementation, that would continue to be sufficient (since there is no difference between a flat directory that exceeds memory limitations vs. a nested directory that excess memory limitations). 

Do we want to wait on this feature until I can figure out how to rework `Dir` into actual async iterating over a directory list?

Reply:
> The current implementation was already reading the entire directory into memory and returning that

This is surprising. I thought `opendir` was created mostly to avoid this.

Reply:
https://github.com/nodejs/node/blob/d4d0b09122d91478881ee7fd36eb3c9f78f8452d/lib/internal/fs/dir.js#L128

The "async" part is just transforming the pieces into `Dirent` on each iteration. I was surprised too!

Reply:
First call to `open`/`read` reads all of the entities into the `Dir` and returns the "first" one. Then every subsequent call pops the next entity off and returns it as a `Dirent`.

Reply:
I would say we should fix the async iteration for opendir first.

Reply:
this has been fixed i believe. please review latest iteration and let me know if you have any similar or additional concerns. Thank you!